Change testdriver.js object parameter to specific parameters.

JuhaVainio opened the pull request:
Earlier APIs took only one object parameter. Now parameters are changed to specific parameters. Tests and documentation also reflect this change.

All debug traces has been also removed.

JuhaVainio committed:
Change testdriver.js object parameter to specific parameters.

Earlier APIs took only one object parameter. Now parameters are changed
to specific parameters. Tests and documentation also reflect this
change.

All debug traces has been also removed.

JuhaVainio commented:
cc @rakuco please check this

rakuco commented on resources/testdriver.js:
```diff
@@ -773,21 +767,17 @@
          *                                to run the call, or null for the current
          *                                browsing context.
          *
          * @returns {Promise} Fulfilled after virtual sensor has been removed,
```
I'd be also say that this is fulfilled if either the virtual sensor has been removed or no virtual sensor of the given type has been created. In other words, this does not reject if create_virtual_sensor() has not been called beforehand.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT_sensors/pull/1/commits/83cc4f75ef04a2699fd59f5ddfe0087a5a509369

rakuco commented on resources/testdriver.js:
```diff
@@ -797,21 +787,17 @@
          *                                to run the call, or null for the current
          *                                browsing context.
          *
          * @returns {Promise} Returns `serialized mock sensor
```
I'd mention that this promise is rejected if there are WebDriver command errors or if a virtual sensor with the given type does not exist.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT_sensors/pull/1/commits/83cc4f75ef04a2699fd59f5ddfe0087a5a509369

rakuco commented on resources/testdriver.js:
```diff
@@ -745,25 +743,21 @@
          *                                browsing context.
          *
          * @returns {Promise} Fulfilled after virtual sensor has been updated,
          * or rejected in the cases the WebDriver command errors.
```
I'd mention that this is also rejected if a virtual sensor of the given type does not exist.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT_sensors/pull/1/commits/83cc4f75ef04a2699fd59f5ddfe0087a5a509369

rakuco commented on resources/testdriver.js:
```diff
@@ -745,25 +743,21 @@
+         * @param {Object} reading - An object for sensor reading values.
+         * Sensor values format depend on used sensor. Used keys are defined in
+         * each sensor specification. Links to sensor specific specs are listed
+         * in above chapter.
```
"Chapter" sounds confusing here, you are just referring to "the text above", not a book chapter or another page :-)

In fact, you could just remove that big list above and just refer to https://w3c.github.io/sensors/#dictdef-mocksensorreadingvalues and mention that each specification defines a different format.

It's also important to note somewhere in the documentation here that will cause the virtual sensor to behave as if a new reading was available from the platform sensor and it may be discarded due to the checks in https://w3c.github.io/sensors/#update-latest-reading.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT_sensors/pull/1/commits/83cc4f75ef04a2699fd59f5ddfe0087a5a509369

rakuco commented on resources/testdriver.js:
```diff
@@ -691,32 +691,30 @@
          *                                browsing context.
          *
          * @returns {Promise} Fulfilled after virtual sensor has been created,
          * or rejected in the cases the WebDriver command errors.
```
I'd also say that this is rejected if a virtual sensor of the same type already exists.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT_sensors/pull/1/commits/83cc4f75ef04a2699fd59f5ddfe0087a5a509369

rakuco commented on resources/testdriver.js:
```diff
@@ -691,32 +691,30 @@
          * "``maxSamplingFrequency``" key holds a double representing frequency
-         * in Hz that is used to set minimum supported sampling frequency for
+         * in Hz that is used to set maximum supported sampling frequency for
          * the associated mock sensor.
```
This is too long, you can just refer to https://w3c.github.io/sensors/#dictionary-mocksensorconfiguration (see e.g. `config` in https://web-platform-tests.org/writing-tests/testdriver.html#test_driver.add_virtual_authenticator or `descriptor` in https://web-platform-tests.org/writing-tests/testdriver.html#test_driver.set_permission).

You should also mention that `sensor_params` is optional, and the defaults will be used if it is not specified.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT_sensors/pull/1/commits/83cc4f75ef04a2699fd59f5ddfe0087a5a509369

rakuco commented on tools/wptrunner/wptrunner/executors/actions.py:
```diff
@@ -281,57 +281,47 @@ class CreateVirtualSensorAction:
         self.protocol = protocol
 
     def __call__(self, payload):
-        print("JV666 > tools/wptrunner/wptrunner/executors/actions.py CreateVirtualSensorAction call()\n")
```
By the way, you can use `self.logger.debug()` to have some debugging messages in all `__call__()`s, just see what the other classes are doing.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT_sensors/pull/1/commits/83cc4f75ef04a2699fd59f5ddfe0087a5a509369

rakuco commented on tools/wptrunner/wptrunner/executors/actions.py:
```diff
@@ -281,57 +281,47 @@ class CreateVirtualSensorAction:
     name = "get_virtual_sensor_information"
 
     def __init__(self, logger, protocol):
-        print("JV666 > tools/wptrunner/wptrunner/executors/actions.py GetVirtualSensorInformation init()\n")
```
Not related do this line, but I couldn't add a comment above: this class should be called "GetVirtualSensorInformation**Action**", all other classes in the file have the same suffix.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT_sensors/pull/1/commits/83cc4f75ef04a2699fd59f5ddfe0087a5a509369

rakuco commented on tools/wptrunner/wptrunner/executors/protocol.py:
```diff
@@ -716,14 +716,17 @@ class VirtualSensorProtocolPart(ProtocolPart):
 
     @abstractmethod
-    def create_virtual_sensor(self, create_parameters):
+    def create_virtual_sensor(self, sensor_type, sensor_params):
```
Also a comment unrelated to this specific line: please add an extra empty line before the class declaration above. See how other classes are separated by two empty lines.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT_sensors/pull/1/commits/83cc4f75ef04a2699fd59f5ddfe0087a5a509369

rakuco commented on tools/wptrunner/wptrunner/executors/executorwebdriver.py:
```diff
@@ -355,25 +355,27 @@ class WebDriverVirtualSensor(VirtualSensorProtocolPart):
+    def get_virtual_sensor_information(self, sensor_type):
+        return self.webdriver.send_session_command("GET", "sensor/%s" % sensor_type)
 
 class WebDriverProtocol(Protocol):
```
Please add an extra empty line before this declaration.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT_sensors/pull/1/commits/83cc4f75ef04a2699fd59f5ddfe0087a5a509369

rakuco commented on tools/wptrunner/wptrunner/executors/executorwebdriver.py:
```diff
@@ -355,25 +355,27 @@ class WebDriverVirtualSensor(VirtualSensorProtocolPart):
-        print("JV666 > tools/wptrunner/wptrunner/executors/executorwebdriver.py get_virtual_sensor_information()\n")
-        type = information_parameters["type"]
-        return self.webdriver.send_session_command("GET", "sensor/%s" % type, information_parameters)
+    def create_virtual_sensor(self, sensor_type, sensor_params):
```
This class should have **Part** at the end of its name like the others.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT_sensors/pull/1/commits/83cc4f75ef04a2699fd59f5ddfe0087a5a509369

rakuco commented on tools/wptrunner/wptrunner/testdriver-extra.js:
```diff
@@ -269,27 +269,19 @@
-        console.trace("JV666 > tools/wptrunner/wptrunner/testdriver-extra.js create_virtual_sensor()");
-        console.trace(JSON.stringify(create_parameters));
-        return create_action("create_virtual_sensor", {create_parameters, context});
+    window.test_driver_internal.create_virtual_sensor = function(sensor_type, sensor_params, context = null) {
```
Style nit: `context=null` rather than `context = null`. `set_spc_transaction_mode` is the outlier in this file.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT_sensors/pull/1/commits/83cc4f75ef04a2699fd59f5ddfe0087a5a509369

rakuco commented on tools/wptrunner/wptrunner/executors/executorwebdriver.py:
```diff
@@ -355,25 +355,27 @@ class WebDriverVirtualSensor(VirtualSensorProtocolPart):
-        type = information_parameters["type"]
-        return self.webdriver.send_session_command("GET", "sensor/%s" % type, information_parameters)
+    def create_virtual_sensor(self, sensor_type, sensor_params):
+        if isinstance(sensor_params, dict):
```
Thinking about this again, I believe the most idiomatic solution is to make sensor_params an optional parameter in the JS bits that defaults to `{}` (an Object) and always assume it's a dict here in Python. You could do something like this:

```js
// testdriver.js
... = create_virtual_sensor(sensor_type, sensor_params={}, context=null) {
  ...
}

// testdriver-extra.js
window.test_driver_internal.create_virtual_sensor = function(sensor_type, sensor_params={}, context=null) {
  ...
}
```

```python
# executorwebdriver.py
def create_virtual_sensor(self, sensor_type, sensor_params):
  body = {"type": sensor_type}
  body.update(sensor_params)
  return self.webddriver.send_session_command(...)
```

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT_sensors/pull/1/commits/83cc4f75ef04a2699fd59f5ddfe0087a5a509369

rakuco reviewed (changes requested):
I've added some comments about issues coming from previous commits here just to use the opportunity to tidy everything up at once.

JuhaVainio committed:
Update testdriver.js documentation

After code review documentation in testdriver.js is updated.

Marked sensor_params-parameter as optional in testdriver.js and in
testdriver-extra.js.

Added traces to actions.py.

Code cleanup.

rakuco commented on resources/testdriver.js:
```diff
@@ -693,68 +693,44 @@
-         * "``maxSamplingFrequency``" key holds a double representing frequency
-         * in Hz that is used to set maximum supported sampling frequency for
-         * the associated mock sensor.
+         * @param {MockSensor} sensor_params - A `MockSensor
```
`MockSensor` is something else, `sensor_params` is a `MockSensorConfiguration` dictionary.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT_sensors/pull/1/commits/f64d4760535ad415bdd01f5c761e5a13aec48885

JuhaVainio committed:
Change MockSensor dictionary to MockSensorConfiguration dictionary

rakuco reviewed (approved):
lgtm!
